cmd/cloud-endpoints-controller/main.go

```diff
@@ -126,84 +130,51 @@ func sync(parent *CloudEndpoint, children *CloudEndpointControllerRequestChildre
+		var finalOpenAPISpec string
 		var err error
 
-		if parent.Spec.OpenAPISpec == nil {
-			// Create default spec
-			if parent.Spec.TargetIngress.Name != "" {
-				// Fetch the ingress
-				ingress, err := config.clientset.ExtensionsV1beta1().Ingresses(parent.Spec.TargetIngress.Namespace).Get(parent.Spec.TargetIngress.Name, metav1.GetOptions{})
-				if err != nil {
-					log.Printf("[INFO][%s] waiting for Ingress %s", parent.Name, parent.Spec.TargetIngress.Name)
-					return status, &desiredChildren, nil
-				}
-
-				// Get target from ingress IP
-				if len(ingress.Status.LoadBalancer.Ingress) < 1 {
-					log.Printf("[INFO][%s] waiting for loadbalancer status from Ingress %s", parent.Name, parent.Spec.TargetIngress.Name)
-					return status, &desiredChildren, nil
-				}
-				target = ingress.Status.LoadBalancer.Ingress[0].IP
-
-				// Populate the jwtAudiences
-				if len(parent.Spec.TargetIngress.JWTServices) > 0 {
-					ingBackends, err := getIngBackends(ingress)
-					if err != nil {
-						return status, &desiredChildren, err
-					}
-					bePatterns := make([]string, len(parent.Spec.TargetIngress.JWTServices))
-
-					for i, svcName := range parent.Spec.TargetIngress.JWTServices {
-						svc, err := config.clientset.CoreV1().Services(parent.Spec.TargetIngress.Namespace).Get(svcName, metav1.GetOptions{})
-						if err != nil {
-							return status, &desiredChildren, fmt.Errorf("Failed to populate JWT audience from kubernetes service, not found: '%s', %v", svcName, err)
-						}
-						if svc.Spec.Type == corev1.ServiceTypeNodePort && len(svc.Spec.Ports) > 0 {
-							nodePort := strconv.Itoa(int(svc.Spec.Ports[0].NodePort))
-							found := false
-							for _, be := range ingBackends {
-								if strings.Contains(be, fmt.Sprintf("k8s-be-%s", nodePort)) {
-									bePatterns[i] = be
-									backend, err := config.clientCompute.BackendServices.Get(config.Project, be).Do()
-									if err == nil {
-										found = true
-										jwtAud := makeJWTAudience(config.ProjectNum, strconv.Itoa(int(backend.Id)))
-										log.Printf("[INFO][%s] Created jwtAud: %s", parent.Name, jwtAud)
-										status.JWTAudiences = append(status.JWTAudiences, jwtAud)
-									}
-								}
-							}
-							if found == false {
-								return status, &desiredChildren, fmt.Errorf("Backend not found or is not ready for service: %s, NodePort: %s", svcName, nodePort)
-							}
-						} else {
-							return status, &desiredChildren, fmt.Errorf("Service %s not type NodePort", svcName)
-						}
-					}
-				}
-			} else {
-				target = parent.Spec.Target
-			}
-			openAPISpec, err = makeDefaultOpenAPISpec(status.Endpoint, target, status.JWTAudiences)
-			if err != nil {
+		if parent.Spec.TargetIngress.Name != "" {
+			target, status.JWTAudiences, err = getTargetIngress(parent)
+			if err != nil { //the ingress isn't ready
+				log.Printf("[INFO][%s] target not ready.", parent.Name)
 				return status, &desiredChildren, err
 			}
```

Comment:
Where is the check to see if target is set?

You need to fix getTargetIngress to only return and error if there is an off-nominal error. The master branch does this already. The correct logic should verify target is not an empty string, if so and if err is nil, then that means it's waiting for the ingress.

Reply:
Because that link returns nil _and_ an empty string, indicating that it's waiting for the ingress, target is not set in this case and the result would be a spec with no target set.